common/cache/metricsScopeCache.go

```diff
@@ -0,0 +1,113 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package cache
+
+import (
+	"bytes"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/common/metrics"
+)
+
+type domainMetricsScopeCache struct {
+	scopeMap map[string]metrics.Scope
+	buffer *buffer
+}
+
+type buffer struct {
+	sync.RWMutex
+	bufferMap map[string]metrics.Scope
+}
+
+var cache atomic.Value
+
+
+// NewDomainMetricsScopeCache constructs a new domainMetricsScopeCache
+func NewDomainMetricsScopeCache() DomainMetricsScopeCache {
+	mc := &domainMetricsScopeCache{
+		scopeMap: make(map[string]metrics.Scope),
+		buffer: &buffer{
+			bufferMap: make(map[string]metrics.Scope),
+		},
+	}
+
+	ticker := time.NewTicker(30 * time.Second)
+	quit := make(chan struct{})
+
+	go func() {
+		for {
+			select {
+			case <- ticker.C:
+				go func() {
+					mc.buffer.Lock()
+					defer mc.buffer.Unlock()
+
+					if len(mc.buffer.bufferMap) > 0 {
+						data := cache.Load().(map[string]metrics.Scope)
```

Comment:
I don't think you can update the underlying map directly. I think this will actually update the map used in Get. I think you need to create a new map and copy everything from buffer and existing map.